src/gfx/egui/renderer.cpp

```diff
@@ -166,62 +168,68 @@ struct EguiRendererImpl {
+  void render(const egui::FullOutput &output, const float4x4 &rootTransform, const gfx::DrawQueuePtr &drawQueue) {
+    meshPool.reset();
+    processPendingTextureFrees();
 
-void EguiRenderer::render(const egui::FullOutput &output, const float4x4 &rootTransform, const gfx::DrawQueuePtr &drawQueue) {
-  impl->meshPool.reset();
-  impl->processPendingTextureFrees();
-
-  // Update textures before render
-  auto &textureUpdates = output.textureUpdates;
-  for (size_t i = 0; i < textureUpdates.numSets; i++) {
-    auto &textureSet = textureUpdates.sets[i];
-    SPDLOG_LOGGER_INFO(logger, "textureSet {}", (uint64_t)textureSet.id);
-    impl->textures.set(textureSet);
-  }
+    // Update textures before render
+    auto &textureUpdates = output.textureUpdates;
+    for (size_t i = 0; i < textureUpdates.numSets; i++) {
+      auto &textureSet = textureUpdates.sets[i];
+      SPDLOG_LOGGER_INFO(logger, "textureSet {}", (uint64_t)textureSet.id);
+      textures.set(textureSet);
+    }
 
-  // Update meshes & generate drawables
-  auto &drawables = impl->drawables;
-  impl->drawables.resize(output.numPrimitives);
-  for (size_t i = 0; i < output.numPrimitives; i++) {
-    auto &prim = output.primitives[i];
-
-    MeshPtr mesh = impl->meshPool.allocateBuffer(prim.numVertices);
-    mesh->update(impl->meshFormat, prim.vertices, prim.numVertices * sizeof(egui::Vertex), prim.indices,
-                 prim.numIndices * sizeof(uint32_t));
-
-    MeshDrawable &drawable = drawables[i];
-    drawable.mesh = mesh;
-    drawable.transform = rootTransform;
-    TexturePtr texture = impl->textures.get(prim.textureId);
-    if (texture) {
-      drawable.parameters.set("color", texture);
-      if (texture->getFormat().pixelFormat == WGPUTextureFormat_R8Unorm) {
-        drawable.parameters.set("flags", uint32_t(0x1));
+    // Update meshes & generate drawables
+    drawables.resize(output.numPrimitives);
+    for (size_t i = 0; i < output.numPrimitives; i++) {
+      auto &prim = output.primitives[i];
+
+      MeshPtr mesh = meshPool.allocateBuffer(prim.numVertices);
+      mesh->update(meshFormat, prim.vertices, prim.numVertices * sizeof(egui::Vertex), prim.indices,
+                   prim.numIndices * sizeof(uint32_t));
+
+      MeshDrawable &drawable = drawables[i];
+      drawable.mesh = mesh;
+      drawable.transform = rootTransform;
+      drawable.features = uiFeatures;
+      TexturePtr texture = textures.get(prim.textureId);
+      if (texture) {
+        drawable.parameters.set("color", texture);
+        if (texture->getFormat().pixelFormat == WGPUTextureFormat_R8Unorm) {
+          drawable.parameters.set("flags", uint32_t(0x1));
+        }
       }
-    }
 
-    drawable.clipRect = int4(prim.clipRect.min.x, prim.clipRect.min.y, prim.clipRect.max.x, prim.clipRect.max.y);
+      // drawable.clipRect = int4(prim.clipRect.min.x, prim.clipRect.min.y, prim.clipRect.max.x, prim.clipRect.max.y);
```

Comment:
TODO: Make this switchable for world space UI, also make it work by either rendering to texture first or using stencil